dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckOwnershipValidationHook.java

```diff
@@ -0,0 +1,235 @@
+            boolean b2 = !user.isAuthorized( Authorities.F_ENROLLMENT_CASCADE_DELETE.getAuthority() );
+            if ( b1
+                && b2 )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1103 )
+                    .addArg( user )
+                    .addArg( pi ) );
+            }
+        }
+
+        // Check acting user is allowed to change/write existing pi and program
+        if ( strategy.isUpdateOrDelete() )
+        {
+            ProgramInstance programInstance = context.getProgramInstance( enrollment.getEnrollment() );
+            trackerImportAccessManager
+                .checkWriteEnrollmentAccess( reporter, programInstance.getProgram(), programInstance );
+        }
+
+        trackerImportAccessManager.checkWriteEnrollmentAccess( reporter, program,
+            new ProgramInstance( program, tei, organisationUnit ) );
+    }
+
+    @Override
+    public void validateEvent( ValidationErrorReporter reporter, Event event )
+    {
+        TrackerImportValidationContext context = reporter.getValidationContext();
+        TrackerImportStrategy strategy = context.getStrategy( event );
+        TrackerBundle bundle = context.getBundle();
+        User user = bundle.getUser();
+
+        checkNotNull( user, USER_CANT_BE_NULL );
+        checkNotNull( event, EVENT_CANT_BE_NULL );
+
+        OrganisationUnit organisationUnit = context.getOrganisationUnit( event.getOrgUnit() );
+        Program program = context.getProgram( event.getProgram() );
+        ProgramStage programStage = context.getProgramStage( event.getProgramStage() );
+        ProgramInstance programInstance = context.getProgramInstance( event.getEnrollment() );
+
+        // Check acting user is allowed to change existing/write event
+        if ( strategy.isUpdateOrDelete() )
+        {
+            validateUpdateAndDeleteEvent( reporter, event, context.getProgramStageInstance( event.getEvent() ) );
+        }
+
+        CategoryOptionCombo categoryOptionCombo = context
+            .getCategoryOptionCombo( event.getAttributeOptionCombo() );
+
+        validateCreateEvent( reporter, user,
+            categoryOptionCombo,
+            programStage,
+            programInstance,
+            organisationUnit,
+            program );
+    }
+
+    protected void validateCreateEvent( ValidationErrorReporter reporter, User actingUser,
+        CategoryOptionCombo categoryOptionCombo, ProgramStage programStage, ProgramInstance programInstance,
+        OrganisationUnit organisationUnit, Program program )
+    {
+        checkNotNull( organisationUnit, ORGANISATION_UNIT_CANT_BE_NULL );
+        checkNotNull( actingUser, USER_CANT_BE_NULL );
+        checkNotNull( program, PROGRAM_CANT_BE_NULL );
+
+        boolean noProgramStageAndProgramIsWithoutReg = programStage == null && program.isWithoutRegistration();
+
+        programStage = noProgramStageAndProgramIsWithoutReg ? program.getProgramStageByStage( 1 ) : programStage;
```

Comment:
This is copied from the existing old code, I'm not sure why this is written like this, to me it looks like when programStage == null && program.isWithoutRegistration() then it will pick program.getProgramStageByStage( 1 ) in all other cases it til pick the incoming/parameter programStage.